app/models/submission/validations_by_template_name.rb

```diff
@@ -65,5 +80,94 @@ def validate_consistent_column_value(column_header)
       )
     end
   end
-  # rubocop:enable Metrics/MethodLength,Metrics/AbcSize
+  # rubocop:enable Metrics/MethodLength
+
+  def validate_samples_per_pool_for_labware
+    return if headers.index(HEADER_BARCODE).nil? && headers.index(HEADER_PLATE_WELLS).nil?
+
+    grouped_rows = group_rows_by_study_and_project
+    grouped_rows.each_value { |rows| process_rows(rows) }
+  end
+
+  private
+
+  def process_rows(rows)
+    barcodes = rows.pluck(headers.index(HEADER_BARCODE))
+    well_locations = rows.pluck(headers.index(HEADER_PLATE_WELLS))
+
+    if plate?(barcodes, well_locations)
+      validate_for_plates(barcodes, well_locations, rows)
+    elsif tube?(barcodes, well_locations)
+      validate_for_tubes(barcodes, rows)
+    end
+  end
+
+  # rubocop:disable Metrics/AbcSize
+  def validate_for_plates(barcodes, well_locations, rows)
+    plate = Plate.find_from_any_barcode(barcodes.uniq.first)
+    return if plate.nil?
+
+    wells = plate.wells.for_bulk_submission.located_at(well_locations)
+    total_number_of_samples_per_study_project = wells.map(&:samples).flatten.count.to_i
+    number_of_pools = rows.pluck(headers.index(HEADER_NUMBER_OF_POOLS)).uniq.first.to_i
+
+    validate_samples_per_pool(rows, total_number_of_samples_per_study_project, number_of_pools)
+  end
+  # rubocop:enable Metrics/AbcSize
+
+  # rubocop:disable Metrics/AbcSize
+  # rubocop:disable Metrics/MethodLength
+  def validate_for_tubes(barcodes, rows)
+    tubes =
+      Receptacle
+        .on_a(Tube)
+        .for_bulk_submission
+        .with_barcode(barcodes)
+        .tap do |found|
+          missing = details['barcode'].reject { |barcode| found.any? { |tube| tube.any_barcode_matching?(barcode) } }
+          if missing.present?
+            raise ActiveRecord::RecordNotFound, "Could not find Tubes with barcodes #{missing.inspect}"
+          end
+        end
+    total_number_of_samples_per_study_project = tubes.map(&:samples).flatten.count.to_i
+    number_of_pools = rows.pluck(headers.index(HEADER_NUMBER_OF_POOLS)).uniq.first.to_i
+
+    validate_samples_per_pool(rows, total_number_of_samples_per_study_project, number_of_pools)
+  end
+  # rubocop:enable Metrics/AbcSize
+  # rubocop:enable Metrics/MethodLength
+
+  def plate?(barcodes, well_locations)
+    barcodes.present? && well_locations.present?
+  end
+
+  def tube?(barcodes, well_locations)
+    barcodes.present? && well_locations.blank?
+  end
+
+  # Checks if the asset is either a tube or a plate.
+  def valid_labware?(barcodes, well_locations)
```

Comment:
I don't like the look of this as a way to validate a labware. Where's this used in the code (search doesn't find it in your PR changes)? and why is there no test?
e.g. a TubeRack is a labware but won't have wells but is not a tube.
Maybe look for a valid sti_type? plate or tube if that's the only 2 types you want to allow?

Reply:
This was probably something I had in a previous iteration and forgot to remove.